src/bin/pg_autoctl/pgsetup.c

```diff
@@ -354,25 +349,56 @@ pg_setup_init(PostgresSetup *pgSetup,
 	}
 
 	/*
-	 * If PostgreSQL is running, register if it's in recovery or not.
+	 * When we have a PGDATA and Postgres is not running, we need to grab more
+	 * information about the local installation: pg_controldata can give us the
+	 * pg-_control_version, catalog_version_no, and system_identifier.
 	 */
-	if (pgSetup->control.pg_control_version > 0 &&
-		pgSetup->pidFile.port > 0 &&
-		pgSetup->pgport == pgSetup->pidFile.port)
+	if (errors == 0)
 	{
 		/*
-		 * Sometimes `pg_ctl start` returns with success and Postgres is still
-		 * in crash recovery replaying WAL files, in the "starting" state
-		 * rather than the "ready" state.
-		 *
-		 * In that case, we wait until Postgres is ready for connections. The
-		 * whole pg_autoctl code is expecting to be able to connect to
-		 * Postgres, so there's no point in returning now and having the next
-		 * connection attempt fail with something like the following:
-		 *
-		 * ERROR Connection to database failed: FATAL: the database system is
-		 * starting up
+		 * Only run pg_controldata when Postgres is not running, otherwise we
+		 * get the same information later from an SQL query, see
+		 * pgsql_get_postgres_metadata.
 		 */
+		if (!pg_setup_is_running(pgSetup) &&
+			pgSetup->control.pg_control_version == 0)
+		{
+			pg_controldata(pgSetup, missing_pgdata_is_ok);
+
+			if (pgSetup->control.pg_control_version == 0)
+			{
+				/* we already logged about it */
+				if (!missing_pgdata_is_ok)
+				{
+					errors++;
+				}
+			}
+
+			log_debug("Found PostgreSQL system %" PRIu64 " at \"%s\", "
+														 "version %u, catalog version %u",
+					  pgSetup->control.system_identifier,
+					  pgSetup->pgdata,
+					  pgSetup->control.pg_control_version,
+					  pgSetup->control.catalog_version_no);
+		}
+	}
+
+	/*
+	 * Sometimes `pg_ctl start` returns with success and Postgres is still in
+	 * crash recovery replaying WAL files, in the "starting" state rather than
+	 * the "ready" state.
+	 *
+	 * In that case, we wait until Postgres is ready for connections. The whole
+	 * pg_autoctl code is expecting to be able to connect to Postgres, so
+	 * there's no point in returning now and having the next connection attempt
+	 * fail with something like the following:
+	 *
+	 * ERROR Connection to database failed: FATAL: the database system is
+	 * starting up
+	 */
+	if (pgSetup->pidFile.port > 0 &&
```

Comment:
As we see a little earlier in the same file:

``` c
		 * Only run pg_controldata when Postgres is not running, otherwise we
		 * get the same information later from an SQL query, see
		 * pgsql_get_postgres_metadata.
```

So yeah we can't rely on `pgSetup->control` to have been updated at this point in the function anymore. It's not a problem though, because we want to connect if Postgres is running. It used to be more like an optimisation: if PGDATA does not exists, no need to check for the PID and stuff.